lms/static/js/edxnotes/utils/notes_collector.js

```diff
@@ -0,0 +1,111 @@
+        var fetchOnce = null,
+            fetchNotes,
+            renderNotes,
+            _requestsData = null,
+            storeNotesRequestData,
+            SearchRequestsData = [],
+            waitUntilAllRequestsStored;
+
+        waitUntilAllRequestsStored = function (maxTimeout) {
+            var deferred = $.Deferred(),
+                intervalId = null,
+                elapsedTimeout = 0,
+                pollInterval = 100,
+                checkRequests;
+
+            checkRequests = function() {
+               if ($('.edx-notes-wrapper').length === SearchRequestsData.length) {
+                   deferred.resolve();
+               } else {
+                   elapsedTimeout += pollInterval;
+                   if (elapsedTimeout >= maxTimeout) {
+                       clearInterval(intervalId);
+                       deferred.reject();
+                   }
+               }
+            };
+
+            intervalId = setInterval(checkRequests, pollInterval);
+            return deferred.promise();
+        };
+
+        storeNotesRequestData = function (data) {
+            SearchRequestsData.push(data);
+            
+            // We only need to call `fetchNotes` once only.
+            if (SearchRequestsData.length === 1) {
+                fetchNotes();                
+            }
+        };
+
+        fetchNotes = function () {
+            console.log('fetchNotes called');
+            waitUntilAllRequestsStored(5000).done(function () {
+                console.log('Fetching notes for all requests...Total Requests: ' + SearchRequestsData.length);
+                // We need this data later for rendering the notes in case of success
+                _requestsData = SearchRequestsData.slice(0);
+
+                // `user` and `course_id` values are same for every annotatable
+                // component so we pick these from first `SearchRequestsData` item
+                var settings = {
+                        data: {
+                            user: SearchRequestsData[0].params.user,
+                            course_id: SearchRequestsData[0].params.courseId
+                        },
+                        type: 'GET',
+                        dataType: "json",
+                        success: renderNotes,
+                        error: Annotator.Plugin.Store.prototype._onError
+                    },
+                    searchEndpoint = SearchRequestsData[0].params.endpoint + 'search/?',
+                    usageIds = _.map(SearchRequestsData, function(item) {
+                        return 'usage_id=' + encodeURIComponent(item.params.usageId);
+                    });
+
+                // Search endpoint expects the below format for query params 
+                // /api/v1/search/?course_id={course_id}&user={user_id}&usage_id={usage_id}&usage_id={usage_id} ...
+                searchEndpoint += usageIds.join('&');
+
+                $.ajax(
```

Comment:
I would use promises on this call, rather than callbacks. This makes it easier to chain dependencies.